qos/evm/check.go

```diff
@@ -50,19 +63,55 @@ func getEndpointCheck(
 }
 
 // withChainIDCheck updates the request context to make an EVM JSON-RPC eth_chainId request.
+// eg. '{"jsonrpc":"2.0","id":1,"method":"eth_chainId"}'
 func withChainIDCheck(requestCtx *requestContext) {
 	requestCtx.jsonrpcReq = buildJSONRPCReq(idChainIDCheck, methodChainID)
 }
 
 // withBlockHeightCheck updates the request context to make an EVM JSON-RPC eth_blockNumber request.
+// eg. '{"jsonrpc":"2.0","id":1,"method":"eth_blockNumber"}'
 func withBlockHeightCheck(requestCtx *requestContext) {
 	requestCtx.jsonrpcReq = buildJSONRPCReq(idBlockNumberCheck, methodBlockNumber)
 }
 
-func buildJSONRPCReq(id int, method jsonrpc.Method) jsonrpc.Request {
-	return jsonrpc.Request{
+// withArchivalBlockCheck updates the request context to make an EVM JSON-RPC eth_getBalance request with a random archival block number.
+// eg. '{"jsonrpc":"2.0","id":1,"method":"eth_getBalance","params":["0x28C6c06298d514Db089934071355E5743bf21d60", "0xe71e1d"]}'
+func withArchivalBlockCheck(requestCtx *requestContext) {
+	// Get the archival check configuration from the endpoint store.
+	// It contains the contract address whose balance is to be checked.
+	archivalCheckConfig := requestCtx.endpointStore.serviceState.archivalCheckConfig
+
+	// Get the current state of the archival check.
+	serviceArchivalState := requestCtx.endpointStore.serviceState.archivalState
+
+	requestCtx.jsonrpcReq = buildJSONRPCReq(
+		idArchivalBlockCheck,
+		methodGetBalance,
+		// Pass params in this order, eg. "params":["0x28C6c06298d514Db089934071355E5743bf21d60", "0xe71e1d"]
+		// Reference: https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance
+		archivalCheckConfig.ContractAddress,
+		serviceArchivalState.blockNumberHex,
+	)
+
+	// Set the archival balance check flag to true.
+	// This is used to ensure that only hydrator requests for the archival block number are used
+	// to update QoS data on whether endpoints are able to service archival requests.
+	requestCtx.archivalBalanceCheck = true
```

Comment:
Could you help clarify why this may be needed?
Suggestion:
- RequestContext should not know about the source of a request (user/hydrator)
- The ArchivalCheck component in the ServiceState can choose to ignore an observation if it is not helpful for establishing Archival State of one or more endpoints.